lib/spack/spack/environment.py

```diff
@@ -1201,7 +1257,10 @@ def install_all(self, args=None):
                 if not spec.package.installed:
                     uninstalled_specs.append(spec)
 
-        for spec in uninstalled_specs:
+        # Sort by total deps so specs have to be installed after dependencies
+        # This ensures dev-builds done in appropriate order
```

Comment:
I think you're going to have to set `use_cache=False` for each dev package explicitly in this loop. The cache flag is used in the `spack.installer.Installer._install_task()` method and currently the `dev-build` command ends up installing the package from the cache instead of the local source if the skip-cache flag isn't set. See this ticket: https://github.com/spack/spack/issues/15340 about that.

Alternatively, you could introduce a whitelist/blacklist of packages to use the cache for and pass that through into the installer class.

Reply:
I put it in a slightly different place in the code, but you're absolutely right I hadn't thought of that.